presto-jdbc/src/main/java/com/facebook/presto/jdbc/ReferenceCountingLoadingCache.java

```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    private final ScheduledExecutorService valueCleanupService;
+    private final LoadingCache<K, Holder<V>> backingCache;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    // TODO: This is probably on the long side. Maybe change to 30 seconds?
+    private static final long retentionPeriod = 2;
+    private static final TimeUnit retentionUnit = TimeUnit.MINUTES;
+
+    private ReferenceCountingLoadingCache(CacheLoader<K, V> loader, Consumer<V> disposer, Ticker ticker)
+    {
+        requireNonNull(loader, "loader is null");
+        requireNonNull(disposer, "disposer is null");
+
+        this.valueCleanupService = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("cache-cleanup-%s"));
+        this.backingCache = CacheBuilder.newBuilder()
+                .removalListener(new RemovalListener<K, Holder<V>>() {
+                    @Override
+                    public void onRemoval(RemovalNotification<K, Holder<V>> notification)
+                    {
+                        Holder<V> holder = notification.getValue();
+                        /*
+                         * The docs say that both the key and value may be
+                         * null if they've already been garbage collected.
+                         * We aren't using weak or soft keys or values, so
+                         * this shouldn't apply.
+                         */
+                        requireNonNull(holder, format("holder is null while removing key %s", notification.getKey()));
+                        checkState(holder.getRefcount() == 0, "Non-zero refcount disposing %s", notification.getKey());
+                        checkState(holder.getCleanerCount() == 0, "Non-zero refcount disposing %s", notification.getKey());
+                        disposer.accept(holder.get());
+                    }
+                })
+                .ticker(ticker)
+                .build(new CacheLoader<K, Holder<V>>() {
+                    @Override
+                    public Holder<V> load(K key)
+                            throws Exception
+                    {
+                        return new Holder<>(loader.load(key));
+                    }
+                });
+    }
+
+    public static class Builder<K, V>
+    {
+        Ticker ticker = Ticker.systemTicker();
+
+        public Builder<K, V> ticker(Ticker ticker)
+        {
+            this.ticker = requireNonNull(ticker, "ticker is null");
+            return this;
+        }
+
+        public ReferenceCountingLoadingCache<K, V> build(CacheLoader<K, V> loader, Consumer<V> disposer)
+        {
+            return new ReferenceCountingLoadingCache<K, V>(loader, disposer, ticker);
+        }
+    }
+
+    public static <K, V> Builder<K, V> builder()
+    {
+        return new Builder<>();
+    }
+
+    public void close()
+    {
+        if (closed.compareAndSet(false, true)) {
+            valueCleanupService.shutdownNow();
```

Comment:
I need to review how this is used. The check could be a good idea, though we should make the method synchronized so no one can sneak in a reference while this is called (assuming that's possible given the broader usage).